src/group_theory/pi_hom.lean

```diff
@@ -0,0 +1,384 @@
+
+@[simp]
+lemma to_fun_empty : to_fun f ∅ = 1 := finset.noncomm_prod_empty _ _
+
+@[simp]
+lemma to_fun_insert_of_not_mem (S : finset I) (i : I) (h : i ∉ S) :
+  to_fun f (insert i S) = ϕ i (f i) * to_fun f S :=
+finset.noncomm_prod_insert_of_not_mem _ _ _ _ h
+
+@[simp]
+lemma to_fun_one : to_fun 1 S = 1 :=
+begin
+   induction S using finset.cons_induction_on with i S hnmem ih,
+   { simp }, { simp [ih, hnmem], }
+end
+
+lemma to_fun_commutes (i : I) (hnmem : i ∉ S) :
+  commute (ϕ i (g i)) (to_fun f S) :=
+begin
+  induction S using finset.induction_on with j S hnmem' ih,
+  { simp, },
+  { simp only [to_fun_insert_of_not_mem _ _ _ _ hnmem'],
+
+    have hij : i ≠ j, by {simp at hnmem, tauto},
+    have hiS : i ∉ S, by {simp at hnmem, tauto},
+    calc ϕ i (g i) * (ϕ j (f j) * (to_fun ϕ f S : G)) -- TODO: Why do I have to mention `ϕ`?
+        = (ϕ i (g i) * ϕ j (f j)) * to_fun ϕ f S : by rw ← mul_assoc
+    ... = (ϕ j (f j) * ϕ i (g i)) * to_fun ϕ f S : by { congr' 1, apply fact.elim hcomm _ _ hij }
+    ... = ϕ j (f j) * (ϕ i (g i) * to_fun ϕ f S) : by rw mul_assoc
+    ... = ϕ j (f j) * (to_fun ϕ f S * ϕ i (g i)) : by { congr' 1, apply ih hiS }
+    ... = (ϕ j (f j) * to_fun ϕ f S) * ϕ i (g i) : by rw ← mul_assoc }
+end
+
+@[simp]
+lemma to_fun_mul (S : finset I) : to_fun (f * g) S = to_fun f S * to_fun g S :=
+begin
+  induction S using finset.induction_on with i S hnmem ih,
+  { simp, },
+  { simp only [to_fun_insert_of_not_mem _ _ _ _ hnmem],
+    rw ih, clear ih,
+    simp only [pi.mul_apply, map_mul],
+    repeat { rw mul_assoc }, congr' 1,
+    repeat { rw ← mul_assoc }, congr' 1,
+    exact (to_fun_commutes _ _ _ S i hnmem), }
+end
+
+lemma to_fun_in_sup_range (S : finset I) :
+  to_fun f S ∈ ⨆ i ∈ S, (ϕ i).range :=
+begin
+  induction S using finset.induction_on with i S hnmem ih,
+  { simp, },
+  { simp only [to_fun_insert_of_not_mem _ _ _ _ hnmem],
+    refine (subgroup.mul_mem _ _ _),
+    { apply subgroup.mem_Sup_of_mem, { use i }, { simp, }, },
+    { refine @bsupr_le_bsupr' _ _ _ _ _ _ (λ i, (ϕ i).range) _ ih,
+      exact λ i, finset.mem_insert_of_mem } }
+end
+
+/-- The canonical homomorphism from a pi group, restricted to a subset  -/
+def hom : (Π (i : I), H i) →* G :=
+{ to_fun := λ f, to_fun f S,
+  map_one' := to_fun_one _,
+  map_mul' := λ f g, to_fun_mul f g S, }
+
+lemma to_fun_update_one (i : I) (y : H i) (S : finset I) :
+  to_fun (function.update 1 i y) S = if i ∈ S then ϕ i y else 1 :=
+begin
+  induction S using finset.induction_on with j S hnmem ih,
+  { simp, },
+  { simp only [ to_fun_insert_of_not_mem _ _ _ _ hnmem],
+    by_cases (i = j),
+    { subst h, rw ih, simp [hnmem], },
+    { change i ≠ j at h,
+      simpa [h, hnmem, function.update_noteq (ne_comm.mp h)] using ih, } }
+end
+
+@[simp]
+lemma hom_update_one (i : I) (y : H i):
+  hom S (function.update 1 i y) = if i ∈ S then ϕ i y else 1 := to_fun_update_one _ _ _
```

Comment:
Should use `mul_single` once https://github.com/leanprover-community/mathlib/pull/11849 lands